python/cugraph/dask/link_analysis/mg_pagerank_wrapper.pyx

```diff
@@ -21,38 +21,78 @@ from cugraph.structure.graph_primtypes cimport *
+    if "value" in input_df.columns:
+        weights = input_df['value']
+        weight_t = weights.dtype
+    else:
+        weight_t = np.dtype("float32")
+
+    # FIXME: Offsets and indices are currently hardcoded to int, but this may
+    #        not be acceptable in the future.
+    numberTypeMap = {np.dtype("int32") : <int>numberTypeEnum.int32Type,
+                     np.dtype("int64") : <int>numberTypeEnum.int64Type,
+                     np.dtype("float32") : <int>numberTypeEnum.floatType,
+                     np.dtype("double") : <int>numberTypeEnum.doubleType}
+
+    # FIXME: needs to be edge_t type not int
+    cdef int num_partition_edges = len(src)
+
+    cdef uintptr_t c_src_vertices = src.__cuda_array_interface__['data'][0]
+    cdef uintptr_t c_dst_vertices = dst.__cuda_array_interface__['data'][0]
+    cdef uintptr_t c_edge_weights = <uintptr_t>NULL
+
+    # FIXME: data is on device, move to host (to_pandas()), convert to np array and access pointer to pass to C
+    cdef uintptr_t c_vertex_partition_offsets = vertex_partition_offsets.values_host.__array_interface__['data'][0]
+
+    cdef graph_container_t graph_container
+
+    populate_graph_container(graph_container,
+                             handle_[0],
+                             <void*>c_src_vertices, <void*>c_dst_vertices, <void*>c_edge_weights,
+                             <void*>c_vertex_partition_offsets,
+                             <numberTypeEnum>(<int>(numberTypeMap[vertex_t])),
+                             <numberTypeEnum>(<int>(numberTypeMap[edge_t])),
+                             <numberTypeEnum>(<int>(numberTypeMap[weight_t])),
+                             num_partition_edges,
+                             num_global_verts, num_global_edges,
+                             partition_row_size, partition_col_size,
+                             True,
+                             True, True) 
 
-    num_verts = local_data['verts'].sum()
-    num_edges = local_data['edges'].sum()
-
-    local_offset = local_data['offsets'][rank]
-    dst = dst - local_offset
-    num_local_verts = local_data['verts'][rank]
-    num_local_edges = len(src)
-
-    cdef uintptr_t c_local_verts = local_data['verts'].__array_interface__['data'][0]
-    cdef uintptr_t c_local_edges = local_data['edges'].__array_interface__['data'][0]
-    cdef uintptr_t c_local_offsets = local_data['offsets'].__array_interface__['data'][0]
-
-    [src, dst] = graph_primtypes_wrapper.datatype_cast([src, dst], [np.int32])
-    _offsets, indices, weights = coo2csr(dst, src, None)
-    offsets = _offsets[:num_local_verts + 1]
-    del _offsets
     df = cudf.DataFrame()
-    df['vertex'] = cudf.Series(np.zeros(num_verts, dtype=np.int32))
-    df['pagerank'] = cudf.Series(np.zeros(num_verts, dtype=np.float32))
+    df['vertex'] = cudf.Series(np.zeros(num_global_verts, dtype=np.int32))
+    df['pagerank'] = cudf.Series(np.zeros(num_global_verts, dtype=np.float32))
```

Comment:
The size of the PageRank vector should be the local number of vertices not the global number of vertices; so the memory footprint per GPU for PageRank data becomes O(V/P).

For the karate data,

GPU0 should have 14 values and GPU1 should have 20 with P_row=1 and P_col = 2

and if I print df['vertex'] after C++ function returns I get 

GPU0
```
0     0.097002
1     0.052878
2     0.031521
3     0.029113
4     0.025639
5     0.024589
6     0.021075
7     0.014309
8     0.014645
9     0.014535
10    0.016785
11    0.014535
12    0.014559
13    0.015043
14    0.000000 <= C++ function does not touch from here
15    0.000000
16    0.000000
17    0.000000
18    0.000000
19    0.000000
20    0.000000
21    0.000000
22    0.000000
23    0.000000
24    0.000000
25    0.000000
26    0.000000
27    0.000000
28    0.000000
29    0.000000
30    0.000000
31    0.000000
32    0.000000
33    0.000000
```

GPU1
```
0     0.100918
1     0.071692
2     0.057078
3     0.035861
4     0.037157
5     0.029765
6     0.029536
7     0.029113
8     0.024491
9     0.026287
10    0.021979
11    0.021979
12    0.019604
13    0.021006
14    0.019573
15    0.014535
16    0.014559
17    0.014535
18    0.014535
19    0.009565
20    0.000000 <= C++ function does not touch from here
21    0.000000
22    0.000000
23    0.000000
24    0.000000
25    0.000000
26    0.000000
27    0.000000
28    0.000000
29    0.000000
30    0.000000
31    0.000000
32    0.000000
33    0.000000
```

and if I print result_pr in test_mg_pagerank.py

I get

```
    pagerank  vertex
0   0.000000       0
1   0.000000       0
2   0.097002       0
3   0.052878       0
4   0.031521       0
5   0.029113       0
6   0.025639       0
7   0.024589       0
8   0.021075       0
9   0.014309       0
10  0.014645       0
11  0.014535       0
12  0.016785       0
13  0.014535       0
14  0.014559       0
15  0.015043       0
16  0.000000       0
17  0.000000       0
18  0.000000       0
19  0.000000       0
20  0.000000       0
21  0.000000       0
22  0.000000       0
23  0.000000       0
24  0.000000       0
25  0.000000       0
26  0.000000       0
27  0.000000       0
28  0.000000       0
29  0.000000       0
30  0.000000       0
31  0.000000       0
32  0.000000       0
33  0.000000       0
```

If this works correctly, the final PageRank values should take the first 14 values from GPU0's result and the first 20 values from GPU1, but the final result does not match with this.

This is a bug.

Reply:
updated

Reply:
Now getting this. Did you forget to push your most recent updates?

```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   df['pagerank'] = cudf.Series(np.zeros(num_local_verts, dtype=weight_t))
E   NameError: name 'num_local_verts' is not defined

cugraph/dask/link_analysis/mg_pagerank_wrapper.pyx:93: NameError
================================================================================================ short test summary info ================================================================================================
FAILED cugraph/tests/dask/test_mg_pagerank.py::test_dask_pagerank[0] - NameError: name 'num_local_verts' is not defined
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=================================================================================================== 1 failed in 9.82s ===================================================================================================
```